File: Source/CSharp/Expressive/Expressive/Operators/Additive/PlusOperator.cs
using Expressive.Expressions;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Expressive.Operators.Additive
{
    internal class PlusOperator : OperatorBase
    {
        #region IOperator Members

        public override string[] Tags { get { return new[] { "+" }; } }

        public override IExpression BuildExpression(Token previousToken, IExpression[] expressions, ExpressiveOptions options)
        {
            if (IsUnary(previousToken))
            {
                return new UnaryExpression(UnaryExpressionType.Plus, expressions[0] ?? expressions[1]);
            }

            return new BinaryExpression(BinaryExpressionType.Add, expressions[0], expressions[1], options);
        }

        public override bool CanGetCaptiveTokens(Token previousToken, Token token, Queue<Token> remainingTokens)
        {
            var remainingTokensCopy = new Queue<Token>(remainingTokens.ToArray());

            return this.GetCaptiveTokens(previousToken, token, remainingTokensCopy).Any();
        }

        public override Token[] GetInnerCaptiveTokens(Token[] allCaptiveTokens)
        {
            return allCaptiveTokens.Skip(1).ToArray();
        }

        public override OperatorPrecedence GetPrecedence(Token previousToken)
        {
            if (IsUnary(previousToken))
            {
                return OperatorPrecedence.UnaryPlus;
            }
            return OperatorPrecedence.Add;
        }

        #endregion

        private bool IsUnary(Token previousToken)
        {
            return string.IsNullOrEmpty(previousToken?.CurrentToken) ||
                string.Equals(previousToken.CurrentToken, "(", StringComparison.Ordinal) ||
                previousToken.CurrentToken.IsArithmeticOperator();
        }
    }
}

File: Source/Java/Expressive/src/main/java/com/bijington/expressive/operators/additive/PlusOperator.java
package com.bijington.expressive.operators.additive;

import com.bijington.expressive.expressions.*;
import com.bijington.expressive.helpers.Strings;
import com.bijington.expressive.operators.OperatorBase;
import com.bijington.expressive.operators.OperatorPrecedence;
import com.sun.org.apache.xpath.internal.operations.Bool;

import java.util.*;

/**
 * Created by shaun on 10/07/2016.
 */
public class PlusOperator extends OperatorBase {
    @Override
    public String[] getTags() { return new String[] { "+" }; }

    @Override
    public IExpression buildExpression(String previousToken, IExpression[] expressions) {
        if (isUnary(previousToken)) {
            return new UnaryExpression(UnaryExpressionType.Plus, (expressions[0] != null) ? expressions[0] : expressions[1]);
        }

        return new BinaryExpression(BinaryExpressionType.Add, expressions[0], expressions[1]);
    }

    @Override
    public Boolean canGetCaptiveTokens(String previousToken, String token, Queue<String> remainingTokens) {
        Queue<String> remainingTokensCopy = new LinkedList<>(remainingTokens);

        return this.getCaptiveTokens(previousToken, token, remainingTokensCopy).length > 0;
    }

    @Override
    public String[] getCaptiveTokens(String previousToken, String token, Queue<String> remainingTokens) {
        String[] result;

        if (isUnary(previousToken)) {
            if (!remainingTokens.isEmpty()) {
                result = new String[] { token, remainingTokens.remove() };
            }
            else {
                result = new String[] { token };
            }
        }
        else {
            result = new String[] { token };
        }

        return result;
    }

    @Override
    public String[] getInnerCaptiveTokens(String[] allCaptiveTokens) {
        return Arrays.copyOfRange(allCaptiveTokens, 1, allCaptiveTokens.length);
    }

    @Override
    public int getPrecedence(String previousToken) {
        if (isUnary(previousToken)) {
            return OperatorPrecedence.UnaryPlus;
        }
        return OperatorPrecedence.Add;
    }

    private Boolean isUnary(String previousToken) {
        return Strings.isNullOrEmpty(previousToken) ||
                previousToken.equals("(") ||
                Strings.isArithmeticOperator(previousToken);
    }
}
